release/mirror: Add examples for using the new mirror

mtnbikenc opened the pull request:
_Resubmission of https://github.com/openshift/origin/pull/23381_

When performing a release mirror, provide examples for use in
install-config.yaml (install) or ImageContentSourcePolcy (upgrade).

https://jira.coreos.com/browse/CORS-1105

Example output:
```
$ oc adm release mirror --from=registry.svc.ci.openshift.org/ocp/release:4.2 --to=registry.svc.ci.openshift.org/rteague-1/mirror-test
info: Mirroring 94 images to registry.svc.ci.openshift.org/rteague-1/mirror-test ...

...
Success
Update image:  registry.svc.ci.openshift.org/rteague-1/mirror-test:release
Mirror prefix: registry.svc.ci.openshift.org/rteague-1/mirror-test

To use the new mirrored repository to install, add the following section to the install-config.yaml:

imageContentSources:
- mirrors:
  - registry.svc.ci.openshift.org/rteague-1/mirror-test
  source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
- mirrors:
  - registry.svc.ci.openshift.org/rteague-1/mirror-test
  source: registry.svc.ci.openshift.org/ocp/release


To use the new mirrored repository for upgrades, use the following to create an ImageContentSourcePolicy:

apiVersion: operator.openshift.io/v1alpha1
kind: ImageContentSourcePolicy
metadata:
  name: example
spec:
  repositoryDigestMirrors:
  - mirrors:
    - registry.svc.ci.openshift.org/rteague-1/mirror-test
    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
  - mirrors:
    - registry.svc.ci.openshift.org/rteague-1/mirror-test
    source: registry.svc.ci.openshift.org/ocp/release
```

mtnbikenc commented:
/cc @smarterclayton

mtnbikenc commented:
/cc @abhinavdahiya

smarterclayton commented:
Temporary hold for https://github.com/openshift/api/pull/384

/hold

Still need to hear the question about what happens on upgrade when you have to create an ICSP instead.

mtrmac commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +450,64 @@ func (o *MirrorOptions) Run() error {
+		Sources []string `json:"sources"`
+	}
+	type installConfigSubsection struct {
+		ImageContentSources []installConfigSubsectionImageContentSources `json:"imageContentSources"`
```
This needs updating for https://github.com/openshift/api/pull/384 .

mtnbikenc replied:
Can you link to an example ICSP?

mtrmac replied:
Not yet, openshift/installer will need to update for that API change first (updating https://github.com/openshift/installer/pull/1910 ) and define the new type.

mtrmac replied:
> Not yet, openshift/installer will need to update for that API change first

This is happening in https://github.com/openshift/installer/pull/2048 .

mtnbikenc replied:
Updated.

mtnbikenc commented:
Refactored for API update.  Updated example in the description.

mtrmac commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +450,67 @@ func (o *MirrorOptions) Run() error {
+	if err != nil {
+		return fmt.Errorf("Unable to marshal repositoryDigestMirrors yaml: %v", err)
+	}
+	fmt.Fprintf(out, "\nTo use the new mirrored repository for upgrades, add the following section to an ImageContentSourcePolicy:\n")
```
Nitpick: Should this mention the `Spec` field of `ImageContentSourcePolicy`?

abhinavdahiya replied:
I think the general goal for the output here is going to be provide information that helps users to take direct action.

So the first part if going to be, how to install using that mirror.
and 
the first part provides sub-section of the install-config because otherwise the whole object is difficult to provide.

The second question is, how to upgrade to that mirror.
and I think since `github.com/openshift/api` is already vendored.. providing an **complete** object that can be pushed to the cluster to enable the upgrade has it's merit.

cc @smarterclayton

smarterclayton replied:
We can always take feedback on this, as long as it's not too hard to do each step.

smarterclayton commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +450,67 @@ func (o *MirrorOptions) Run() error {
+	repositories[sourceRepo] = struct{}{}
+
+	for repository := range repositories {
+		sources = append(sources, imageContentSource{
```
Sort these so they display in the same order every time.

smarterclayton commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +450,67 @@ func (o *MirrorOptions) Run() error {
+	if err != nil {
+		return fmt.Errorf("Unable to marshal repositoryDigestMirrors yaml: %v", err)
+	}
+	fmt.Fprintf(out, "\nTo use the new mirrored repository for upgrades, add the following section to an ImageContentSourcePolicy:\n")
```
Add an extra line of whitespace between text and yaml (and up above)

mtnbikenc commented:
@smarterclayton @abhinavdahiya
Refactored to use types defined in API and generate complete ICSP.  Sorted sources by Source.  Added newlines.  Updated example in the description.

abhinavdahiya commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +451,65 @@ func (o *MirrorOptions) Run() error {
+	fmt.Fprintf(out, string(installConfigExample))
+
+	// Create and display ImageContentSourcePolicy example
+	icsp := operatorv1alpha1.ImageContentSourcePolicy{
```
Should include a name maybe?

smarterclayton commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,5 +451,62 @@ func (o *MirrorOptions) Run() error {
+			RepositoryDigestMirrors: sources,
+		},
+	}
+	icspExample, err := yaml.Marshal(icsp)
```
This unfortunately creates creationTimestamp nil.  You might want to convert to unstructured, then remove the creationTimestamp field, then marshal to yaml.

abhinavdahiya commented:
/hold cancel

mtnbikenc commented:
@smarterclayton @abhinavdahiya
For ICSP, added `apiVersion`, `kind` and `name`.  Hacked out `creationTimestamp`.
Updated example in the description.

abhinavdahiya commented:
/lgtm

abhinavdahiya commented:
ping @smarterclayton @mfojtik

smarterclayton commented on pkg/cli/admin/release/mirror.go:
```diff
@@ -442,18 +451,71 @@ func (o *MirrorOptions) Run() error {
+// printImageContentInstructions provides exapmles to the user for using the new repository mirror
+// https://github.com/openshift/installer/blob/master/docs/dev/alternative_release_image_sources.md
+func printImageContentInstructions(out io.Writer, from, to string, repositories map[string]struct{}) error {
+	type installConfigSubsection struct {
```
Is it possible this gets called with no entries?

mtnbikenc replied:
If `repositories` had no entries it would mean the release image didn't reference any images.  In this case, only one item would be printed for mirror usage, which is provided by `--from`.

```
imageContentSources:
- mirrors:
  - registry.svc.ci.openshift.org/rteague-1/mirror-test
  source: quay.io/openshift-release-dev/ocp-release
```

mtnbikenc commented:
Fixed an issue with error handling and used unstructured to remove `creationTimestamp`.

If this looks good I can squash.

abhinavdahiya commented:
ping @mfojtik @deads2k @smarterclayton

smarterclayton commented:
Looks good, once this merges can you add an e2e test in openshift/origin PR that

1. grabs the current release value (from cluster version)
2. runs `oc adm release mirror ...` in a pod (see the tests for images like test/extended/images/info.go for an example of what code you can use) and mirrors to an image stream inside the registry (use the internal registry spec, an image stream in your namespace)
3. check the result for success and the output to match these values

smarterclayton commented:
Please squash

mtnbikenc commented:
Squashed.  All green.

abhinavdahiya commented:
ping @smarterclayton @deads2k @mfojtik

smarterclayton commented:
/lgtm
